### ISSUE TITLE
How to add depot visits based on condition?

### ISSUE THREAD

**Issue by YooTimmy** [google/or-tools#1445]
First of all, thank you for this great library, it really helped a lot!

Now I am trying to resolve a VRP with some special constrains for depot visit. Here are the details:

Imagine that there is a truck company that handles import & export transactions between its customers(e.g. small manufacturing factories ) and port. The import(port -> customer) & export (customer -> port) jobs all has a requested time window to fulfill. There are a number of vehicles available to handle these tasks, and also they will have a max operation time per day(e.g. 10 hours). The schedule is planned on a daily basis. And the objective for the planning to achieve the delivery with minimum number of vehicles.

Until now it looks like a typical VRPTW problem. However, there are some additional request: 

1) When import & export transactions are combined, the truck does not need to visit depot to pick up empty containers. i.e. the truck is able to re-use the empty container from previous import or export transaction.
2) When import & import transactions are combined, the truck needs to visit depot once after the first import transaction to drop off the empty container.
3) When export & export transactions are combined, the truck needs to visit depot before each export transaction to pick up an empty container for usage.

In other words, the vehicle would need to visit a specific depot under certain conditions I have mentioned. So far I was not able to find a similar case in the forum. Anyone has some suggestions or advice? Appreciated the help in advance!

**Comment by jmarca** [comment]
This is tricky, and will take thinking.  With the caveat that I *haven't* done the heavy thinking that is required...here are my suggestions.

First, I just replied to your other question about the need to have a unique node for each event.  So keep that in mind.  If the truck needs to visit a depot to drop off an empty container, that depot drop off node has to be a unique node in the model, even though it shares a location with all the other depot nodes.

So what I would do is think about this in terms of the containers.  Suppose a location A has both import and export transactions that need to happen.  `Ai` and `Ae` maybe.  So if I understand correctly, a valid sequence would be 

```
P1 -> Ai
Ai -> Ae
Ae -> P2
```

Note P1 and P2 are different logical nodes.  P1 and Ai are a pickup and dropoff pair as are Ae and P2.

If you have a location with two import transactions, `B1i` and `B2i`, then you have to do

```
P1 -> B1i
B1i -> (drop empty container1)
(d. e. c.1) -> P2
P2 -> B2i
B2i -> (drop empty container2)
```

So that is the difficulty in figuring this out.  You will have to generate *enough* dummy nodes to drop off the extra empty containers. 

In terms of how to set it up, I'd set up a container dimension, with the "demand" callback function reflecting whether a container is generated or taken away.  For import transactions, the port node generates a container, the destination node does not (callback returns 0), and the (drop empty) node takes the container away (-1), resetting the truck load to 0.  A node that both imports and exports, you can have its demand callback for the import transaction be -1 (takes away a container to be used later) and the export transaction node's demand callback returns +1 (gives the truck a container), and I guess the exporting port has a corresponding -1.  I think simple balancing of sums should show about how many dummy nodes for handling removing empty containers you'll need.

Another possible wrinkle would be managing the pool of empty containers, in case trucks need to pickup an empty to bring to an export transaction.  you might need to keep track of that too if there is a possibility that you'll run out of containers.  I don't know enough about the details of the problem.

Hope that helps

**Comment by YooTimmy** [comment]
Thank you for your time to review the problem! I will take some time to digest your suggestions first.
